### PR TITLE
Link for github issues in PR

### PR DESCRIPTION
**Pull request by CTY-git**
## PR Checklist

- [x] The commit message follows our guidelines: [Code of conduct](https://github.com/patched-codes/patchwork/blob/main/CODE_OF_CONDUCT.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Does this PR introduce a breaking change?
- [x] Include PR in release notes?

## PR Type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Build /CI
- [ ] Documentation
- [ ] Others

## What is the current behavior?
Currently we don't link issues to PRs.

Issue Number: N/A

## What is the new behavior?

When provided with `issue_url`, the `CreatePR` step now will link the newly created PR with the issue.

## Other information

### CODE REVIEW

**Comment by patched-admin** [comment]
This pull request review examines several aspects of a proposed change to add an `issue_url` parameter to the `create_pr` function, noting potential bugs if the `scm_client.create_pr` method does not support this new parameter or requires a non-null value, and highlighting the possibility of creating a new PR if no existing PRs are found. A notable security concern is the need to validate and sanitize the `issue_url` to prevent SSRF attacks. While the code aligns generally with existing coding standards, inconsistencies in parameter naming conventions are pointed out, which may cause confusion. Additionally, another modification in the pull request involves a version bump in the pyproject.toml file from 0.0.84 to 0.0.85. For this change, no bugs or security vulnerabilities are anticipated, and it adheres to coding standards by correctly updating the version in the specified format. Overall, the addition is seen as correctly integrated, provided due diligence is maintained regarding input sanitation and parameter compatibility.

------

<div markdown="1">

* File changed: [patchwork/steps/CreatePR/CreatePR.py](https://github.com/patched-codes/patchwork/pull/1122/files#diff-6f7d9cad8d3bb8c511281e360b5038717dc6590c402e597ab408aec045a6d5d4)
  1. **Potential Bugs:**
   - The added `issue_url` parameter in the `create_pr` function is set to default to None. If the `scm_client.create_pr` method does not support `issue_url` or if it requires a non-null value, this could result in a bug or runtime error.
   - If `scm_client.find_prs` returns an empty iterator for `prs`, `next(iter(prs), None)` will return `None`, which is currently handled. Ensure that this expectation is intended, as this will lead to a new PR creation.

2. **Security Vulnerabilities:**
   - Ensure that `issue_url` is properly validated and sanitized before being used or stored. URL inputs can often be a vector for attacks such as SSRF (Server-Side Request Forgery) if not properly handled.

3. **Coding Standards:**
   - The code change seems to follow the general coding style of the surrounding code. However, ensure consistent use of parameter naming conventions, such as `original_branch` and `feature_branch`, which might differ from `base_branch_name` and `target_branch_name`. This discrepancy might lead to confusion or inconsistencies in future code usage.

Overall, the addition of the `issue_url` parameter appears correctly integrated into the existing create_pr logic, but attention should be paid to input sanitation and parameter compatibility with the used API/library methods.

</div>

<div markdown="1">

* File changed: [patchwork/steps/CreatePR/typed.py](https://github.com/patched-codes/patchwork/pull/1122/files#diff-27b139e8d7ae3df11f774d449f6a94be0f14cf4fd728b421bfaebd2ad0a2f8cd)
  1. **Potential Bugs**: No apparent bugs are introduced by this change since it only adds a new configuration parameter `issue_url` to the `CreatePRInputs` class and follows the existing pattern of configuration annotations.

2. **Security Vulnerabilities**: Introducing additional URL fields, like `issue_url`, could potentially open up security vulnerabilities if input validation is not handled correctly elsewhere in the codebase. Ensure that all input handling involving URLs is properly sanitized to mitigate against injection attacks.

3. **Coding Standards Adherence**: The addition of `issue_url` seems to adhere to existing coding standards by matching the Annotated pattern seen in the surrounding code. However, it's important to ensure consistency in naming conventions as 'issue_url' aligns with preceding field patterns like 'scm_url'.

</div>

<div markdown="1">

* File changed: [pyproject.toml](https://github.com/patched-codes/patchwork/pull/1122/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711)
  The modification in this pull request involves only a version bump from 0.0.84 to 0.0.85 in the pyproject.toml file. 
1. **Potential Bugs**: Since there is only a version number change, there should be no bugs introduced by this modification.
2. **Security Vulnerabilities**: The update does not introduce any new code that could lead to security vulnerabilities. However, it is important to ensure that other dependencies are also up-to-date for enhanced security.
3. **Adherence to Coding Standards**: The modification adheres to the original coding standards as it correctly updates the version number in the specified format within the pyproject.toml file.

</div>

**Comment by patched-admin** [comment]
The pull request review identifies several potential issues and improvements in the code changes. There are potential bugs related to the new parameters `issue` and `issue_url` introduced in the `create_pull` and `create_pr` methods, as these additions may not be supported by their respective libraries and may cause unexpected behavior if not handled properly. Though no immediate security vulnerabilities were noted, it is advised to validate and sanitize the `issue_url` to prevent injection attacks. In terms of coding standards, the changes largely adhere to existing patterns, but there's a noted inconsistency in naming conventions within the `create_pr` function parameters, suggesting a need for uniformity across the project. Furthermore, the use of `logger.warn` should be updated to `logger.warning` to meet modern standards. The addition of default values for new parameters like `issue_url` helps maintain backward compatibility, and while a version bump from 0.0.84 to 0.0.85 is noted, the review emphasizes ensuring that related files and documentation reflect these changes appropriately.

------

<div markdown="1">

* File changed: [patchwork/common/client/scm.py](https://github.com/patched-codes/patchwork/pull/1122/files#diff-a4a0213f577e66fc33444faad6c1991a8c60deb8fc7d5b5527069a66f59804f5)
  1. **Potential Bugs**:
   - The `create_pull` method call now includes an additional argument `issue=issue_obj` which may not be supported by default. This change needs to be validated if the library expects this parameter and manages it correctly.

2. **Security Vulnerabilities**:
   - There are no immediate security vulnerabilities visible with the given changes. However, care should be taken with `issue_url` to ensure it's sanitized and validated before use to prevent any injection attacks.

3. **Coding Standards**:
   - The refactoring to use a helper method (`__find_issue_by_id`) improves code readability and adheres to good abstraction practices.
   - In `create_pr`, an unused parameter `issue_url` is documented as such for gitlab, which is a good practice, but ideally, compile-time checks or different method signatures for different platforms could prevent runtime errors.
   - Logging uses `logger.warn`, which is deprecated in Python's logging library in favor of `logger.warning`. This should be updated to adhere to modern standards.

</div>

<div markdown="1">

* File changed: [patchwork/steps/CreatePR/CreatePR.py](https://github.com/patched-codes/patchwork/pull/1122/files#diff-6f7d9cad8d3bb8c511281e360b5038717dc6590c402e597ab408aec045a6d5d4)
  1. **Potential Bugs:**
   - The call to `scm_client.create_pr` now includes a new `issue_url` parameter. It is important to ensure that the `scm_client.create_pr` method implementation supports this parameter. Otherwise, it may raise an error or unexpected behavior may occur.

2. **Security Vulnerabilities:**
   - If `issue_url` is incorporated without proper validation or sanitization, there could be a potential risk of injection attacks or other vulnerabilities depending on how it is used downstream. Make sure to validate and sanitize the `issue_url` input appropriately.

3. **Coding Standards:**
   - Ensure consistency in naming conventions and style used across the file/project. For example, the existing code uses `base_branch_name` and `target_branch_name` rather than `original_branch` and `feature_branch`, respectively, in the `create_pr` function. The newly introduced parameters should be consistent with the existing code style and naming conventions.

4. **Backward Compatibility:**
   - Adding a default value for `issue_url` as `None` in the `create_pr` function ensures backward compatibility. This seems to be correctly handled in the current changes.

</div>

<div markdown="1">

* File changed: [patchwork/steps/CreatePR/typed.py](https://github.com/patched-codes/patchwork/pull/1122/files#diff-27b139e8d7ae3df11f774d449f6a94be0f14cf4fd728b421bfaebd2ad0a2f8cd)
  1. Potential Bugs: The addition of 'issue_url' seems straightforward as a new configuration item. However, it should be ensured that all parts of the code that deal with 'issue_url' handle it safely, particularly because it is a URL and may require validation of its format.

2. Security Vulnerabilities: Storing any form of URL can introduce security vulnerabilities if not validated correctly. There needs to be input validation to ensure that the 'issue_url' is not malicious (e.g., cross-site scripting attacks or other injection attacks).

3. Coding Standards: The addition maintains the existing pattern used in the class for other attributes. Thus, it adheres to the original coding standards of the module.

</div>

<div markdown="1">

* File changed: [pyproject.toml](https://github.com/patched-codes/patchwork/pull/1122/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711)
  The provided diff indicates a version bump from 0.0.84 to 0.0.85 in the pyproject.toml file for the 'patchwork-cli' project. Since only the version number is changed, and no actual code modifications are present, we cannot identify potential bugs, security vulnerabilities, or adherence to coding standards based solely on this diff. Ensure that other related files, tests, or documentation have been appropriately updated to reflect and support the version bump, if applicable.

</div>

**Comment by patched-admin** [comment]
The pull request review highlights a number of potential issues, security vulnerabilities, and coding standards adherence in the code changes. Notable potential bugs include the risk of inadequate error handling when `find_issue_by_data` returns `None` and the need to ensure that `scm_client.create_pr` can handle the new `issue_url` parameter without causing runtime errors. Security vulnerabilities are not directly introduced, but caution is advised to validate the URLs properly to guard against injection attacks and to check how `issue_url` might affect API requests, particularly concerning sensitive data. In terms of coding standards, the review praises the consistent application of type hints, the maintenance of backward compatibility, and adherence to good coding practices such as separation of concerns. However, it notes the absence of sufficient documentation for the `issue_url` parameter, which could confuse future contributors. Additionally, the changes maintain a consistent coding style and structure, particularly with existing configurations like `Annotated` and `StepTypeConfig`. Lastly, a version update in the `pyproject.toml` file, resulting in no new bugs or security concerns, reflects standard version increment protocols.

------

<div markdown="1">

* File changed: [patchwork/common/client/scm.py](https://github.com/patched-codes/patchwork/pull/1122/files#diff-a4a0213f577e66fc33444faad6c1991a8c60deb8fc7d5b5527069a66f59804f5)
  1. **Potential Bugs:**
   - In the `find_issue_by_url` method, there is a change in the way the URL is parsed. The new code assumes that `get_slug_and_id_from_url` could return `None`, which is a good safety check.
   - When calling `__find_issue_by_id`, if an issue is not found (`None`), the `find_issue_by_data` quietly returns `None`. There's a potential for missing detailed error logging or handling, which could help diagnose issues in production.

2. **Security Vulnerabilities:**
   - The code seems to handle exceptions related to GitHub issues retrieval properly with a generic `GithubException`. However, consider adding more specific exception handling to cover different reasons a request might fail. Also, ensure no sensitive information is exposed in log messages.

3. **Coding Standards:**
   - The new parameter `issue_url` has been consistently added to several `create_pr` functions, preserving backward compatibility by providing a default `None` value, which is good.
   - The addition of `issue_url` as an unused parameter to the GitLab-related `create_pr` method is noted with a comment, suggesting adherence to a consistent API across the SCM implementations, which aligns with good coding standards.
   - The private method `__find_issue_by_id` encapsulates the repository interaction logic, improving code readability and separation of concerns, adhering to good coding practices.
   - The code correctly uses type hints consistently across its changes, maintaining the standard guidelines.

Overall, the changes seem well-structured and consistent with previous code regarding style and functionality, with added robustness for issue handling.

</div>

<div markdown="1">

* File changed: [patchwork/steps/CreatePR/CreatePR.py](https://github.com/patched-codes/patchwork/pull/1122/files#diff-6f7d9cad8d3bb8c511281e360b5038717dc6590c402e597ab408aec045a6d5d4)
  1. **Potential Bugs**: The addition of the `issue_url` parameter to the `create_pr` function doesn't seem to be supported in the call to `scm_client.create_pr`. Ensure that `scm_client.create_pr` can handle the `issue_url` argument. If not, this could cause a runtime error or unintended behavior.

2. **Security Vulnerabilities**: Adding `issue_url` does not inherently introduce new security vulnerabilities, but ensure that URLs are properly validated to prevent potential injection attacks or misinterpretations of the URL within the system.

3. **Coding Standards**: The code seems to follow general Python coding standards. Ensure that the `issue_url` follows the same naming conventions and documentation practices as other similar parameters across the codebase. No explicit docstring modification or comment explaining the newly injected `issue_url` parameter might lead to some confusion for future maintainers or contributors.

</div>

<div markdown="1">

* File changed: [patchwork/steps/CreatePR/typed.py](https://github.com/patched-codes/patchwork/pull/1122/files#diff-27b139e8d7ae3df11f774d449f6a94be0f14cf4fd728b421bfaebd2ad0a2f8cd)
  1. **Security Vulnerability:**
   - The addition of an `issue_url` as a configuration input is generally safe; however, there is a possible security vulnerability if the `issue_url` can be specified or is exposed to untrusted sources. Ensure that the input is properly validated and sanitized to prevent URL injection attacks or other security issues related to URL handling.
   - Consider whether sensitive data is being appended to `issue_url` or if it’s included in any API requests without verification, as this could lead to credential leakage, especially if API keys are involved in URL results or are logged alongside the URL.

2. **Coding Standards:**
   - The change follows the existing coding patterns such as using `Annotated` and `StepTypeConfig`. No standards issues are apparent in terms of deviating from the file's current organization or style.

3. **Potential Bugs:**
   - There might be a potential bug if the `issue_url` is not being used properly later in the code. Ensure that adding this field doesn’t influence logic elsewhere unless intentionally revisited in separate parts of the code. Verify integration with existing URL handling and related logic to avoid null pointer or type errors if `issue_url` is being used wrongfully.

</div>

<div markdown="1">

* File changed: [pyproject.toml](https://github.com/patched-codes/patchwork/pull/1122/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711)
  The provided diff only updates the version number in the `pyproject.toml` file from `0.0.84` to `0.0.85`. There are no potential bugs, security vulnerabilities, or deviations from coding standards introduced by this update. The modification is standard procedure for version incrementing and adheres to the original coding standards.

</div>

File: patchwork/steps/CreatePR/CreatePR.py [hunk 1]
```diff
@@ -147,17 +149,19 @@ def create_pr(
     base_branch_name: str,
     target_branch_name: str,
     scm_client: ScmPlatformClientProtocol,
+    issue_url: str = None,
```
Comment (jonahdc) [inline patchwork/steps/CreatePR/CreatePR.py:152]: Would this be `issue_url: str | None = None`?

**Review by jonahdc — APPROVED** [review]

File: patchwork/steps/CreatePR/CreatePR.py (hunk 1 above)
Comment (CTY-git) [reply patchwork/steps/CreatePR/CreatePR.py]: yep, good catch.

**Comment by patched-admin** [comment]
The pull request review highlights concerns across three main areas: security vulnerabilities, adherence to coding standards, and potential bugs. There is a potential security risk identified with hardcoded URLs in the workflow file, which might necessitate using parameters or secrets to mitigate unauthorized actions or exposure of sensitive information, while also ensuring that exception logging does not inadvertently expose sensitive details. It is recommended to validate URLs to avoid errors like 404s, address potential unpacking issues with the `resource_slug_and_id`, and revise the logging strategy to mask sensitive data. Coding standards are moderately adhered to with clear parameter names, though inconsistencies in naming conventions (e.g., branch-related parameters) are noted, and the use of Python 3.10 syntax might cause compatibility issues with earlier versions. Ensuring all new variables are documented and that appropriate error handling and unit tests are added for new functionalities are also advised. Additionally, the review notes metadata changes, such as version updates, which do not appear to introduce any functional risks. The overall recommendation is to ensure consistency, proper exception handling, validation, and that potential security concerns are adequately addressed.

------

<div markdown="1">

* File changed: [.github/workflows/test.yml](https://github.com/patched-codes/patchwork/pull/1122/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88)
  1. **Security Vulnerability**: Including hardcoded issue URLs directly in the workflow file might create a security concern if the URL triggers any unauthorized actions, or if the issue URL is sensitive or confidential. Consider passing this as a parameter or using a secret if it's sensitive.

2. **Adherence to Coding Standards**: Ensure that the inclusion of the `--issue_url` parameter is consistent with other optional parameters' use within the workflow. This includes spacing, alignment, and any commenting standards used to describe the purpose of new flags or parameters.

3. **Potential Bugs**: It's essential to validate that the provided URLs are reachable or correct where applicable, as hardcoding attributes like URL could lead to 404 errors or incorrect issue associations if even a minor typo is made.

</div>

<div markdown="1">

* File changed: [patchwork/common/client/scm.py](https://github.com/patched-codes/patchwork/pull/1122/files#diff-a4a0213f577e66fc33444faad6c1991a8c60deb8fc7d5b5527069a66f59804f5)
  1. **Potential Bugs:**
   - In the `create_pr` method for GitHub, when `issue_obj` is determined and `issue_url` is non-null, the method `self.get_slug_and_id_from_url(issue_url)` is called. There is no validation performed post call to ensure if `slug` and `issue_id` are correctly set from the tuple. If there is any issue within `get_slug_and_id_from_url`, it would lead to `TypeError` as the tuple will not unpack properly. This can cause issues if `resource_slug_and_id` is `None`.

2. **Security Vulnerabilities:**
   - Logging the exception `e` directly in `find_issue_by_id` method without sanitization or masking can unintentionally expose sensitive information within the logs. It is critical to ensure that any output logs are adequately scrubbed of sensitive repository or issue details.

3. **Adherence to Coding Standards:**
   - Usage of `|` in type annotations like `issue_url: str | None = None` is Python 3.10+ syntax. Make sure this does not break compatibility if the project supports older Python versions.
   - The use of `NotSet` from `GithubObject` to denote a non-existing issue is a good practice; however, this should be implemented consistently across all methods that are handling GitHub issues.
   - There's no immediate indication of test functions being updated or tested for these scenarios. Ensure that any call paths using these functions have corresponding unit tests to evaluate the new code paths and parameter handling.

Overall, it's recommended to double-check the handling of `resource_slug_and_id` post assignment to prevent potential `None` dereferencing and ensure proper exception handling for logging.

</div>

<div markdown="1">

* File changed: [patchwork/steps/CreatePR/CreatePR.py](https://github.com/patched-codes/patchwork/pull/1122/files#diff-6f7d9cad8d3bb8c511281e360b5038717dc6590c402e597ab408aec045a6d5d4)
  1. **Potential Bugs**: 
   - The `typing_extensions.Optional` import might be unnecessary if just using `Optional`. Instead, use `from typing import Optional` unless there's a specific reason to use `typing_extensions`.

2. **Security Vulnerabilities**: 
   - There is no obvious introduction of security vulnerabilities, but it's important to ensure that any URLs or inputs passed from user input are properly validated and sanitized to prevent issues like injection attacks (even though they may not appear in this snippet, considerations should be made in the context).

3. **Code Standards**: 
   - The code appears to adhere to basic Python conventions with clear naming and consistent parameter passing. 
   - However, there's a lack of consistency in naming for parameters compared to their usage in functions (`original_branch` vs `base_branch`, `feature_branch` vs `target_branch`). It might be beneficial to unify these terms for clarity unless they have distinct meanings in different contexts. 
   - Ensure all newly introduced variables (like `issue_url`) are documented if there is a documentation standard in place for the project.
   - There isn't a change in exception handling around the `create_pr` call. If this operation can fail, considering adding appropriate error handling or logging if not done elsewhere in the project.

</div>

<div markdown="1">

* File changed: [patchwork/steps/CreatePR/typed.py](https://github.com/patched-codes/patchwork/pull/1122/files#diff-27b139e8d7ae3df11f774d449f6a94be0f14cf4fd728b421bfaebd2ad0a2f8cd)
  1. **Potential Bugs**: The addition of the `issue_url: Annotated[str, StepTypeConfig(is_config=True)]` field looks straightforward; however, there's no indication of how this new field is being utilized within the rest of the codebase. It's important to check if this field needs initialization or default validation to prevent any runtime errors.

2. **Security Vulnerabilities**: Adding a new string for a URL shouldn't inherently introduce a security vulnerability. However, if this URL is later used for making web requests or integrations, ensure that proper validation and sanitization are in place to prevent potential vulnerabilities such as injection attacks or request forgeries.

3. **Coding Standards**: There's no explicit violation of coding standards from this diff alone, as it follows the existing pattern for other URL fields like `scm_url`. However, please ensure consistency in the documentation and usage across the codebase related to this new field.

</div>

<div markdown="1">

* File changed: [pyproject.toml](https://github.com/patched-codes/patchwork/pull/1122/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711)
  The change in the version number from 0.0.84 to 0.0.85 in the pyproject.toml file appears to be a standard update to indicate a new version of the software. There doesn't seem to be any potential bugs or security vulnerabilities introduced by this change as it only modifies metadata.

Since this is a metadata change, it does not involve functional code, so it does not break any original coding standards in regard to code logic. However, ensure that any corresponding changes that depend on this version update (e.g., in changelogs or documentation) are also addressed outside of this diff.

Overall, the modification maintains coding standards in terms of proper semantic versioning.

</div>

**Comment by CTY-git** [comment]
Not required, original method of adding `fixes {issue_url}` works when relevant permission is set.

**Pull request closed without merging**